### PR TITLE
Merge HTTP resource restructure with master branch

### PR DESCRIPTION
**Pull request by chamil321**

### CODE REVIEW

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/request.bal [hunk 1]
```diff
@@ -1,91 +1,292 @@
...
+@Param {value:"req: The inbound request message"}
+@Return {value:"The map of form params"}
+public function <InRequest request> getFormParams () (map) {
+    mime:Entity entity = request.getEntity();
+    string formData = mime:getText(entity);
+    map parameters = {};
+    if (formData != null && formData != "") {
+        string[] entries = formData.split("&");
+        int i = 0;
+        while (i < lengthof entries) {
```
Comment (shafreenAnfar) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/request.bal]: Is it possible to give a proper name for this "i" variable.

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java [hunk 2]
```diff
@@ -248,28 +248,16 @@ private void validateResourceSignature(HttpResource resource) {
...
-        if (reqParamDetail == null) {
-            throw new BallerinaConnectorException("request parameter cannot be null");
-        }
-        if (reqParamDetail.getVarType().getPackagePath() == null
-                || !reqParamDetail.getVarType().getPackagePath().equals(Constants.PROTOCOL_PACKAGE_HTTP)
-                || !reqParamDetail.getVarType().getName().equals(Constants.REQUEST)) {
-            throw new BallerinaConnectorException("request parameter should be of type - "
-                                                          + Constants.PROTOCOL_PACKAGE_HTTP + ":" + Constants.REQUEST);
+        //validate connection parameter
+        if (isValidParam(paramDetails.get(0), Constants.CONNECTION)) {
```
Comment (shafreenAnfar) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java]: Shall we remove the comment and change the function name to isValidResourceParam ?

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java [hunk 3]
```diff
@@ -248,28 +248,16 @@ private void validateResourceSignature(HttpResource resource) {
...
+            throw new BallerinaConnectorException("first parameter should be of type - "
+                    + Constants.PROTOCOL_PACKAGE_HTTP + ":" + Constants.CONNECTION);
```
Comment (pubudu91) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java]: The exception should be thrown if the param is not valid right? i.e: this should read `if (!isValidParam)`

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java [hunk 4]
```diff
@@ -248,28 +248,16 @@ private void validateResourceSignature(HttpResource resource) {
...
-        }
-        if (respParamDetail.getVarType().getPackagePath() == null
-                || !respParamDetail.getVarType().getPackagePath().equals(Constants.PROTOCOL_PACKAGE_HTTP)
-                || !respParamDetail.getVarType().getName().equals(Constants.RESPONSE)) {
-            throw new BallerinaConnectorException("response parameter should be of type - "
-                                                          + Constants.PROTOCOL_PACKAGE_HTTP + ":" + Constants.REQUEST);
+        //Validate request parameter
+        if (isValidParam(paramDetails.get(1), Constants.IN_REQUEST)) {
+            throw new BallerinaConnectorException("second parameter should be of type - "
+                    + Constants.PROTOCOL_PACKAGE_HTTP + ":" + Constants.IN_REQUEST);
```
Comment (pubudu91) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java]: Same comment as above

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/request.bal (hunk 1 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/request.bal]: Ack

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java (hunk 2 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java]: Ack

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java (hunk 3 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HTTPServicesRegistry.java]: Ack

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal [hunk 5]
```diff
@@ -1,218 +1,195 @@
...
+
+@Description { value:"Sends response to the caller"}
+@Param { value:"conn: The server connector connection" }
+@Param { value:"res: A outbound response message" }
+@Return { value:"Error occured during HTTP server connector respond" }
+public native function <Connection conn> respond (OutResponse res) (HttpConnectorError);
+
+@Description { value:"Forwards response to the caller"}
+@Param { value:"conn: The server connector connection" }
+@Param { value:"res: A inbound response message" }
```
[the remark this answers is not in the document]
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: Ack

File: samples/ballerina-by-example/examples/outbound-request-with-multiparts/outbound-request-with-multiparts.description [hunk 6]
```diff
@@ -0,0 +1,3 @@
+//With Ballerina http client connector, you can send multiparts via outbound requests. First build the body parts
```
Comment (pubudu91) [inline samples/ballerina-by-example/examples/outbound-request-with-multiparts/outbound-request-with-multiparts.description]: HTTP*

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal [hunk 7]
```diff
@@ -1,218 +1,195 @@
...
+
+@Description {value:"Represent 'Content-Legth' header name"}
+public const string CONTENT_LENGTH = "Content-Length";
+
+@Description { value:"Represents the HTTP server connector connection"}
+@Field {value:"remoteHost: The server host name"}
 @Field {value:"port: The server port"}
+public struct Connection {
+	string remoteHost;
+	int port;
```
[the remark this answers is not in the document]
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: These two attributes are suitable and make sense to the context of a connection.

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/Constants.java [hunk 8]
```diff
@@ -184,21 +183,29 @@
...
-    public static final int REQUEST_HEADERS_INDEX = 0;
-
-    //Response struct indexes
-    public static final int RESPONSE_STATUS_CODE_INDEX = 0;
-    public static final int RESPONSE_REASON_PHRASE_INDEX = 0;
-    public static final int RESPONSE_SERVER_INDEX = 1;
-    public static final int RESPONSE_HEADERS_INDEX = 0;
+    public static final int HTTP_MESSAGE_INDEX = 0;
+    public static final int ENTITY_INDEX = 1;
+    public static final int ENTITY_BODY_REQUIRED_INDEX = 0;
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/Constants.java]: Do we need the unused variable ENTITY_BODY_REQUIRED_INDEX?

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Forward.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        BStruct inboundResponseStruct = (BStruct) getRefArgument(context, 1);
+        HTTPCarbonMessage requestMessage = HttpUtil.getCarbonMsg(connectionStruct, null);
+
+        HttpUtil.checkFunctionValidity(connectionStruct, requestMessage);
+        HTTPCarbonMessage responseMessage = HttpUtil
+                .getCarbonMsg(inboundResponseStruct, HttpUtil.createHttpCarbonMessage(false));
+
+        return HttpUtil.prepareResponseAndSend(context, this, requestMessage, responseMessage,
+                inboundResponseStruct);
+    }
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Forward.java]: Shall we move the above code block to HttpUtil class or any other common place as it is duplicated in Respond.java class as well?

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/request/GetProperty.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.natives.AbstractNativeFunction;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+import org.ballerinalang.natives.annotations.ReturnType;
+import org.ballerinalang.net.http.HttpUtil;
+
+/**
+ * Ballerina function to set a message property.
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/request/GetProperty.java]: Incorrect description. It should be changed as get a message property?

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/Constants.java (hunk 8 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/Constants.java]: Removed

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/response/GetProperty.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.natives.AbstractNativeFunction;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+import org.ballerinalang.natives.annotations.ReturnType;
+import org.ballerinalang.net.http.HttpUtil;
+
+/**
+ * Ballerina function to set a message property.
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/response/GetProperty.java]: Incorrect description. It should be changed as get a message property? Check all GetPropertye classes

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Forward.java (hunk 9 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Forward.java]: Created a abstract class and moved it there

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/response/GetProperty.java (hunk 11 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/inbound/response/GetProperty.java]: Ack

**Pull request merged**